test/source_test.go

```diff
@@ -1,20 +1,338 @@
+		llb.AddEnv("HOST", host),
+		llb.AddEnv("PORT", fmt.Sprintf("%d", port)),
+	).Root()
+
+	// tell git to use the credential helper for the host:port combination
+	worker = worker.Run(
+		llb.Shlex(`sh -c 'git config --global credential."http://${HOST}:${PORT}.helper" "/usr/local/bin/frontend credential-helper --kind=token"'`),
+		llb.AddEnv("HOST", host),
+		llb.AddEnv("PORT", fmt.Sprintf("%d", port)),
+	).Root()
+	return worker
+}
+
+func getGitServerAddrs(ctx context.Context, t *testing.T, c gwclient.Client) (string, string) {
+	const (
+		rootlessOutsideAddr = "localhost"
+		rootlessInsideAddr  = "10.0.2.2" // as per the docs
+	)
+
+	if !isRootless(ctx, t, c) {
+		addr := getExtraHostRootful(t)
+		return addr, addr
+	}
+
+	// These may fast-fail the test if the configuration is wrong
+	dockerdPid := getDockerdPid(t)
+	assertDockerdEnvironment(t, dockerdPid)
+
+	return rootlessOutsideAddr, rootlessInsideAddr
+}
+
+func nullCharSplit(data []byte, atEOF bool) (int, []byte, error) {
+	var (
+		i   int
+		err error
+	)
+
+	for i = 0; i < len(data); i++ {
+		if data[i] == 0 {
+			break
+		}
+	}
+
+	if atEOF {
+		err = bufio.ErrFinalToken
+	}
+
+	return i + 1, data[:i], err
+}
+
+func assertDockerdEnvironment(t *testing.T, dockerdPid int) {
+	const (
+		envVarRootlessKitLocalhostKey   = "DOCKERD_ROOTLESS_ROOTLESSKIT_DISABLE_HOST_LOOPBACK"
+		envVarRootlessKitLocalhostValue = "false"
+	)
+
+	f, err := os.Open(fmt.Sprintf("/proc/%d/environ", dockerdPid))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	scanner := bufio.NewScanner(f)
+	scanner.Split(nullCharSplit)
+
+	for scanner.Scan() {
+		k, v, ok := strings.Cut(scanner.Text(), "=")
+		if !ok || k != envVarRootlessKitLocalhostKey {
+			continue
+		}
+
+		if v != envVarRootlessKitLocalhostValue {
+			t.Fatalf("You have a rootless setup. In order to permit TCP service forwarding, you must restart the docker daemon with the %q env var set to %q", envVarRootlessKitLocalhostKey, envVarRootlessKitLocalhostValue)
+		}
+	}
+}
+
+func getDockerdPid(t *testing.T) int {
+	const dockerProcName = "dockerd"
+	procs, err := ps.Processes()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	currentUser, err := user.Current()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var pid int
+
+outer:
+	for _, proc := range procs {
+		name := proc.Executable()
+		if filepath.Base(name) != dockerProcName {
+			continue
+		}
+
+		pp := proc.Pid()
+
+		f, err := os.Open(fmt.Sprintf("/proc/%d/status", pp))
+		if err != nil {
+			t.Fatal(err)
+		}
+		scanner := bufio.NewScanner(f)
+
+		for scanner.Scan() {
+			t := scanner.Text()
+			_, val, ok := strings.Cut(t, "Uid:")
+			if !ok {
+				continue
+			}
+
+			val = strings.TrimSpace(val)
+			if !strings.HasPrefix(val, currentUser.Uid) {
+				continue outer
+			}
+
+			pid = pp
+			break outer
+		}
+	}
+
+	if pid == 0 {
+		t.Fatal("cannot find pid of docker daemon")
+	}
+	return pid
+}
+
+func runGitServer(addr string, port int) {
```

Comment:
Can you pass through a `*testing.T` shutdown the listener in a `t.Cleanup` function?